evm/p2p/peer.py

```diff
@@ -280,48 +338,117 @@ def match_protocols(self, remote_capabilities):
             offset += proto_klass.cmd_length
 
 
+class LESPeer(BasePeer):
+    _les_proto = None
+    _supported_sub_protocols = [LESProtocol]
+
+    def __init__(self, remote, privkey, reader, writer, aes_secret, mac_secret,
+                 egress_mac, ingress_mac, chaindb, network_id):
+        super(LESPeer, self).__init__(
+            remote, privkey, reader, writer, aes_secret, mac_secret,
+            egress_mac, ingress_mac, chaindb, network_id)
+        self._header_fetching_lock = asyncio.Lock()
+
+    @asyncio.coroutine
+    def fetch_headers(self, block_number, reorg_depth=0):
+        """Fetch all headers from our canonical chain head up to block_number.
+
+        If reorg_depth is provided, fetch headers from (current_head - reorg_depth) up to
+        block_number.
+        """
+        with (yield from self._header_fetching_lock):
+            self.logger.info("fetch_headers({}, reorg_depth={}) called".format(
+                block_number, reorg_depth))
+            head_number = self.chaindb.get_canonical_head().block_number
+            head_number -= reorg_depth
+            announced_head_number = block_number
+            while announced_head_number > head_number:
+                request_id = gen_request_id()
+                target_head = min(head_number + MAX_HEADERS_FETCH, announced_head_number)
+                self.les_proto.send_get_block_headers(
+                    target_head, MAX_HEADERS_FETCH, request_id, reverse=True)
+                reply = yield from self.wait_for_reply(request_id)
+                for header in reversed(reply['headers']):
+                    self.chaindb.persist_header_to_db(header)
+                    # FIXME: The reference to the canonical chain head should not be stored in the
+                    # chaindb as that is shared by all peers -- instead, that should be kept on a
+                    # peer-specific db.
```

Comment:
Let me see if I'm following this correctly.  The idea is that we'd sync the chain for each peer, and then we'd merge the various peer-specific chains into a single canonical chain?

Reply:
Yeah, my current thinking is that we can have multiple peers fetching headers and storing them in a shared (the main) DB, but the info about the canonical chain will have to be specific to every peer and in the LightChain/LightClient we'll need some way to reconcile the info from the various peers and provide a consistent view